### ISSUE TITLE
Transaction sign fails

### ISSUE THREAD

**Issue by ArtemKo7v** [bitpay/bitcore#1199]
```
var transaction = new bitcore.Transaction();
transaction.fromString('010000000141ea295a987931d291bfdc3e36cba9172347b50e239c0a40c9ab88d84d9e2d530300000047522102c6d69432088f0f51019b58fa8c9108b2f752f9085849c0fb0b7421e17d0fda472102efe635ea1e7893bd1b6f5ab4ddf59e49d7dfbb0efec95e24e1ade3ac39bf31dc52aeffffffff0336150000000000001976a91436d371a9d4156430d56b9ac72b50193b38d30d2e88ac781e0000000000006951210257f98a9e29f57b18adbce629141cee942bf85eb2cb45be79b5462c04591d31ad2103ef48ae943ae61c44cdb5a79b4d2585002523efea4cf0f23b831985893b62db192102c6d69432088f0f51019b58fa8c9108b2f752f9085849c0fb0b7421e17d0fda4753aeca0404000000000047522102c6d69432088f0f51019b58fa8c9108b2f752f9085849c0fb0b7421e17d0fda472102efe635ea1e7893bd1b6f5ab4ddf59e49d7dfbb0efec95e24e1ade3ac39bf31dc52ae00000000');
var key = bitcore.PrivateKey.fromWIF('xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'); // This was a real PK WIF
transaction.sign(key);
```

Javascript throws an exception:
i {message: "Invalid state: undefined", stack: "Invalid state: undefined↵Error↵    at Error.i (fil…ject.InjectedScript.evaluate (<anonymous>:682:21)", name: "bitcore.ErrorInvalidState"}

```
tansaction.toJSON();

{"version":1,"inputs":[{"prevTxId":"532d9e4dd888abc9400a9c230eb5472317a9cb363edcbf91d23179985a29ea41","outputIndex":3,"sequenceNumber":4294967295,"script":"OP_2 33 0x02c6d69432088f0f51019b58fa8c9108b2f752f9085849c0fb0b7421e17d0fda47 33 0x02efe635ea1e7893bd1b6f5ab4ddf59e49d7dfbb0efec95e24e1ade3ac39bf31dc OP_2 OP_CHECKMULTISIG"}],"outputs":[{"satoshis":5430,"script":"OP_DUP OP_HASH160 20 0x36d371a9d4156430d56b9ac72b50193b38d30d2e OP_EQUALVERIFY OP_CHECKSIG"},{"satoshis":7800,"script":"OP_1 33 0x0257f98a9e29f57b18adbce629141cee942bf85eb2cb45be79b5462c04591d31ad 33 0x03ef48ae943ae61c44cdb5a79b4d2585002523efea4cf0f23b831985893b62db19 33 0x02c6d69432088f0f51019b58fa8c9108b2f752f9085849c0fb0b7421e17d0fda47 OP_3 OP_CHECKMULTISIG"},{"satoshis":263370,"script":"OP_2 33 0x02c6d69432088f0f51019b58fa8c9108b2f752f9085849c0fb0b7421e17d0fda47 33 0x02efe635ea1e7893bd1b6f5ab4ddf59e49d7dfbb0efec95e24e1ade3ac39bf31dc OP_2 OP_CHECKMULTISIG"}],"nLockTime":0}
```

**Comment by maraoz** [comment]
Hey @ArtemKo7v . Can you copy the full exception trace? It's hard to help with so little information about what went wrong

**Comment by shayanb** [comment]
the trace is something like the following:
```
Invalid state: undefined
Error
    at Error.NodeError (/node_modules/bitcore/lib/errors/index.js:20:41)
    at Object.module.exports.checkState (/node_modules/bitcore/lib/util/preconditions.js:9:13)
    at PublicKeyHashInput.getSignatures (/node_modules/bitcore/lib/transaction/input/publickeyhash.js:35:5)
    at forEachInput (/node_modules/bitcore/lib/transaction/transaction.js:907:18)
    at Function.forEach (/node_modules/bitcore/node_modules/lodash/dist/lodash.js:3297:15)
    at Transaction.getSignatures (/node_modules/bitcore/lib/transaction/transaction.js:906:5)
    at Transaction.sign (/node_modules/bitcore/lib/transaction/transaction.js:894:15)
    at Object.signTransaction (/lib/wallet.js:130:15)
    at /test/test_cloud_wallet.js:25:33
    at Request._callback (/lib/wallet.js:24:7)
```
The issue is bitcore.transaction adds a json key to inputs as "output" that has almost the same values as the outputs but includes the fee in the "satoshi", or something like this, can't remember the exact difference. and this "output" is not imported when you use transaction.fromString or even new transaction(TX_HEX).

**Comment by ArtemKo7v** [comment]
```
transaction.sign(key);

bitcore.js:9432 Uncaught NodeError {message: "Invalid state: undefined", stack: "Invalid state: undefined↵Error↵    at Error.NodeEr…ject.InjectedScript.evaluate (<anonymous>:682:21)", name: "bitcore.ErrorInvalidState"} 

module.exports.checkState @ bitcore.js:9432
Transaction.sign @ bitcore.js:8441
(anonymous function) @ VM63:2
InjectedScript._evaluateOn @ VM54:883
InjectedScript._evaluateAndWrap @ VM54:816InjectedScript.evaluate @ VM54:682
```

**Comment by hudgins** [comment]
Here's an example you can paste into the Node repl. Because the "output" property disappears, any attempt to sign the transaction fails the validation.

```
var bitcore = require('bitcore');
var tx = new bitcore.Transaction().from({"address":"mkVuZV2kVtMzddQabFDanFi6DTwWYtgiCn","txid":"65f662321479d9256fca407c0363bb754a6f9321f9851160ed79e4134e4e579c","vout":1,"scriptPubKey":"76a91436a5ee46338acf885538ebd709a810b361c93a4388ac","amount":0.48386244});
tx.toJSON()
// '{"version":1,"inputs":[{"prevTxId":"65f662321479d9256fca407c0363bb754a6f9321f9851160ed79e4134e4e579c","outputIndex":1,"sequenceNumber":4294967295,"script":"","output":{"satoshis":48386244,"script":"OP_DUP OP_HASH160 20 0x36a5ee46338acf885538ebd709a810b361c93a43 OP_EQUALVERIFY OP_CHECKSIG"}}],"outputs":[],"nLockTime":0}'

var tx2 = new bitcore.Transaction(tx.toString())
tx2.toJSON()
// '{"version":1,"inputs":[{"prevTxId":"65f662321479d9256fca407c0363bb754a6f9321f9851160ed79e4134e4e579c","outputIndex":1,"sequenceNumber":4294967295,"script":""}],"outputs":[],"nLockTime":0}'
```

**Comment by braydonf** [comment]
When using `transaction.from` the previous output is added to the input, however when the transaction is serialized/deserialized the previous output is lost (other than a reference from prevTxId and outputIndex).

For a PublicKeyHashInput instance, the method `input.getSignatures` uses that output to determine if the input previous output script can be signed by the supplied private key.

Using transaction.toObject/fromObject may be a better option for serialization to keep all the necessary properties.

**Comment by hudgins** [comment]
My example was contrived in an effort to localize it to bitcore. The actual reason this is an issue for us is because we receive the serialized transaction from another source.

**Comment by braydonf** [comment]
Manually adding the signature to the input may be the only option in that case (however, without access to the previous output info, there is no way to know if it's correct).

**Comment by Christewart** [comment]
Just to note I ran into this issue as well, and I am unclear of how the 'toObject()' method properly serializes the transaction. When trying to transfer it over the wire via and input the string version of this is [object Object]. I've also tried converting the transaction to json using 'toJSON' which seems to work. However, when I try and deserialize the transaction back into native bitcore objects I get an invalid argument error. I've tried using both the construction and the 'fromJSON' method on the 'Transaction' object. 

caught bitcore.ErrorInvalidArgument: Invalid Argumentmodule.exports.checkArgument @ bitcore.js:9875BN.fromNumber @ bitcore.js:1103Object.defineProperty.set @ bitcore.js:7791Output._fromObject @ bitcore.js:7798Output @ bitcore.js:7760(anonymous function) @ bitcore.js:8399forEach @ bitcore.js:38372Transaction.fromObject @ bitcore.js:8398Transaction.fromJSON @ bitcore.js:8348handleGamePayoutUtxos @ suredbits.gamePayout.js:59(anonymous function) @ suredbits.wallet.js:189(anonymous function) @ bitcore-explorers.min.js:1u @ bitcore-explorers.min.js:1n @ bitcore-explorers.min.js:1

and here is the the JSON that I've got from my transaction. 

{"version":1,"inputs":[{"prevTxId":"9df007e38ed5ebc91693351a6b93af8cc27f4f60e824a4a20856cb9cd6b76d02","outputIndex":0,"sequenceNumber":4294967295,"script":"OP_0 72 0x3045022100aef4eba402a2d8c058f441ad8397b9f628dc654e5e24fc44eadd88c112c9ff8b02205de2a9295e286a49dd1b1a813edfa44c559d2ff6599516bd6971f2c7fa4a11a301 OP_PUSHDATA1 105 0x5221025e9adcc3d65c11346c8a6069d6ebf5b51b348d1d6dc4b95e67480c34dc0bc75c21030e7ce7e2694dbfa2bfb3f740c080f0389d59570724bcdd92c60b335984e4db7721030e7ce7e2694dbfa2bfb3f740c080f0389d59570724bcdd92c60b335984e4db7753ae","output":{"satoshis":100000,"script":"OP_HASH160 20 0xe30caeacaebfafe47433ed65c8977339560ccccf OP_EQUAL"},"threshold":2,"publicKeys":["025e9adcc3d65c11346c8a6069d6ebf5b51b348d1d6dc4b95e67480c34dc0bc75c","030e7ce7e2694dbfa2bfb3f740c080f0389d59570724bcdd92c60b335984e4db77","030e7ce7e2694dbfa2bfb3f740c080f0389d59570724bcdd92c60b335984e4db77"],"signatures":[null,null,{"publicKey":"030e7ce7e2694dbfa2bfb3f740c080f0389d59570724bcdd92c60b335984e4db77","prevTxId":"9df007e38ed5ebc91693351a6b93af8cc27f4f60e824a4a20856cb9cd6b76d02","outputIndex":0,"inputIndex":0,"signature":"3045022100aef4eba402a2d8c058f441ad8397b9f628dc654e5e24fc44eadd88c112c9ff8b02205de2a9295e286a49dd1b1a813edfa44c559d2ff6599516bd6971f2c7fa4a11a3","sigtype":1}]},{"prevTxId":"8e417e242e1521a4b521afe4eca7f805acb7fa570cb26ade6722495839b1cdc7","outputIndex":0,"sequenceNumber":4294967295,"script":"OP_0 72 0x3045022100a46327002114fed04461312e6864e6d46dc8bc079c218caa4fd9605cc866da8002207a26bab20b9fe7aaf607e368727d8748bd57ad060be80e54e32125dddb606f8801 OP_PUSHDATA1 105 0x5221025e9adcc3d65c11346c8a6069d6ebf5b51b348d1d6dc4b95e67480c34dc0bc75c21030e7ce7e2694dbfa2bfb3f740c080f0389d59570724bcdd92c60b335984e4db7721030e7ce7e2694dbfa2bfb3f740c080f0389d59570724bcdd92c60b335984e4db7753ae","output":{"satoshis":100000,"script":"OP_HASH160 20 0xe30caeacaebfafe47433ed65c8977339560ccccf OP_EQUAL"},"threshold":2,"publicKeys":["025e9adcc3d65c11346c8a6069d6ebf5b51b348d1d6dc4b95e67480c34dc0bc75c","030e7ce7e2694dbfa2bfb3f740c080f0389d59570724bcdd92c60b335984e4db77","030e7ce7e2694dbfa2bfb3f740c080f0389d59570724bcdd92c60b335984e4db77"],"signatures":[null,null,{"publicKey":"030e7ce7e2694dbfa2bfb3f740c080f0389d59570724bcdd92c60b335984e4db77","prevTxId":"8e417e242e1521a4b521afe4eca7f805acb7fa570cb26ade6722495839b1cdc7","outputIndex":0,"inputIndex":1,"signature":"3045022100a46327002114fed04461312e6864e6d46dc8bc079c218caa4fd9605cc866da8002207a26bab20b9fe7aaf607e368727d8748bd57ad060be80e54e32125dddb606f88","sigtype":1}]}],"outputs":[{"satoshis":null,"script":"OP_DUP OP_HASH160 20 0xaa63d2c040819fdbb0020f6c8cb132692fc4b62e OP_EQUALVERIFY OP_CHECKSIG"}],"nLockTime":0,"changeScript":"OP_DUP OP_HASH160 20 0x4d043a7cfce2dfe5fccdecc7d582da60c9758a9d OP_EQUALVERIFY OP_CHECKSIG","fee":{"amount":0.0001,"code":"BTC"}};
